source/iterator/iterator.go

```diff
@@ -0,0 +1,255 @@
+		})
+		if err != nil {
+			return nil, fmt.Errorf("new shapshot iterator: %w", err)
+		}
+	} else {
+		it.cdc, err = newCDCIterator(
+			ctx,
+			cdcParams{
+				db:            it.db,
+				table:         it.table,
+				trackingTable: it.trackingTable,
+				keys:          it.keys,
+				batchSize:     it.batchSize,
+				columnTypes:   it.tableInfo.ColumnTypes,
+				position:      pos,
+			},
+		)
+		if err != nil {
+			return nil, fmt.Errorf("new cdc iterator: %w", err)
+		}
+	}
+
+	return it, nil
+}
+
+// HasNext returns a bool indicating whether the iterator has the next record to return or not.
+// If the underlying snapshot iterator returns false, the combined iterator will try to switch to the cdc iterator.
+func (c *CombinedIterator) HasNext(ctx context.Context) (bool, error) {
+	switch {
+	case c.snapshot != nil:
+		hasNext, err := c.snapshot.HasNext(ctx)
+		if err != nil {
+			return false, fmt.Errorf("snapshot has next: %w", err)
+		}
+
+		if !hasNext {
+			if er := c.switchToCDCIterator(ctx); er != nil {
+				return false, fmt.Errorf("switch to cdc iterator: %w", err)
+			}
+
+			return c.cdc.HasNext(ctx)
+		}
+
+		return true, nil
+
+	case c.cdc != nil:
+		return c.cdc.HasNext(ctx)
+
+	default:
+		return false, nil
+	}
+}
+
+// Next returns the next record.
+func (c *CombinedIterator) Next(ctx context.Context) (sdk.Record, error) {
+	switch {
+	case c.snapshot != nil:
+		return c.snapshot.Next(ctx)
+
+	case c.cdc != nil:
+		return c.cdc.Next(ctx)
+
+	default:
+		return sdk.Record{}, ErrNoInitializedIterator
+	}
+}
+
+// Stop the underlying iterators.
+func (c *CombinedIterator) Stop() error {
+	if c.snapshot != nil {
+		return c.snapshot.Stop()
+	}
+
+	if c.cdc != nil {
+		return c.cdc.Stop()
+	}
+
+	return nil
+}
+
+// Ack check if record with position was recorded.
+func (c *CombinedIterator) Ack(ctx context.Context, rp sdk.Position) error {
+	pos, err := position.ParseSDKPosition(rp)
+	if err != nil {
+		return fmt.Errorf("parse position: %w", err)
+	}
+
+	if pos.IteratorType == position.TypeCDC {
+		return c.cdc.Ack(ctx, pos)
+	}
+
+	return nil
+}
+
+func (c *CombinedIterator) switchToCDCIterator(ctx context.Context) error {
+	err := c.snapshot.CloseRows()
+	if err != nil {
+		return fmt.Errorf("close rows: %w", err)
+	}
+
+	c.snapshot = nil
+
+	c.cdc, err = newCDCIterator(
+		ctx,
+		cdcParams{
+			db:            c.db,
+			table:         c.table,
+			trackingTable: c.trackingTable,
+			keys:          c.keys,
+			batchSize:     c.batchSize,
+			columnTypes:   c.tableInfo.ColumnTypes,
+			position:      nil,
+		},
+	)
+	if err != nil {
+		return fmt.Errorf("new cdc iterator: %w", err)
+	}
+
+	return nil
+}
+
+func (c *CombinedIterator) setKeys(cfgKeys []string) {
+	// first priority keys from config.
+	if len(cfgKeys) > 0 {
+		for i := range cfgKeys {
+			cfgKeys[i] = strings.ToUpper(cfgKeys[i])
+		}
+
+		c.keys = cfgKeys
+
+		return
+	}
+
+	// second priority primary keys from table.
+	if len(c.keys) > 0 {
+		return
```

Comment:
Where exactly do we set the keys to be primary keys from the table?

Reply:
I decided to move this logic to here, it will be more clear